### ISSUE TITLE
Expose unauthenticated AES-CTR and ChaCha20 ciphers

### ISSUE THREAD

**Issue by ghedo** [briansmith/ring#693]
The current IETF QUIC specification defines a packet number obfuscation scheme that uses either AES-CTR or ChaCha20 (depending on the negotiated TLS 1.3 cipher) "to protect the packet number from casual observation" https://tools.ietf.org/html/draft-ietf-quic-tls-14#section-5.3

It would be great if ring exposed those ciphers, with functions like `insecure_unauthenticated_encrypt_in_place` and `insecure_unauthenticated_decrypt_in_place` (and related `*Key` types) or something along those lines, to allow QUIC implementations written in Rust to use ring for all cryptographic needs without having to resort to ugly hacks like using the OpenSSL EVP API, or using other Rust crypto libraries of dubious provenance just to implement packet number obfuscation.

In #642 someone also requested exposing ChaCha20, but for QUIC we really also need AES-CTR.

**Comment by ghedo** [comment]
FTR, for QUIC implementation use, it would probably be enough to expose the `AES-CTR()` and `ChaCha20()` functions directly rather than having to build a whole API.

**Comment by briansmith** [comment]
Sure, I agree that we should expose the AES-CTR and ChaCha20 mechanisms if QUIC is going to use them like this, so that we can support the implementation of QUIC. Note that the APIs don't necessarily need to be "in place" like the current AEAD API; I think it would be useful to hear from QUIC implementers about what API they'd want. I think it would be good for *ring* to maintain (store and increment) the counter state internally and opaquely to the caller if the QUIC protocol allows such an implementation.

**Comment by ghedo** [comment]
I agree that the API doesn't need to be in-place since the QUIC packet tnumber is going to be at most 4 bytes in all cases (so copying isn't super expensive).

Also, the QUIC packet number obfuscation scheme uses an explicit nonce for both AES and ChaCha (which is basically 16 bytes extracted from the packet's encrypted payload section), so there's no need to keep track of the counter because it's going to be provided by the API.

For example, in both my C and Rust (intended to replace the C one, but still incomplete) implementations I'm actually using the `CRYPTO_chacha_20()` function for encryption/decryption of the packet number (BoringSSL doesn't offer a `EVP_chacha20()`), and I simply extract the first 4 bytes of the explicit nonce to be used as the counter argument.

AES is slightly more annoying because it requires initialization, but an API like the SealingKey/OpeningKey one would work well for both AES and ChaCha I think. The application would simply keep track of the *Key structs rather than the raw [u8] keys (like I'm currently doing for the QUIC packet encryption, which uses a proper AEAD).

**Comment by djc** [comment]
@ghedo thanks for opening this, I was just looking at this yesterday. It would be great if we could have these implementations in *ring* rather than having to rely on other crates for these primitives.

**Comment by djc** [comment]
@briansmith if you can provide some implementation guidance I might be able to work on this.

**Comment by briansmith** [comment]
@djc Look at the AES-GCM code and notice how there are *many* implementations of AES-GCM in *ring* depending on what assembly language modules are available. In particular, there are some where AES-GCM is integrated into one thing, some where there is a special AES-CTR implementation, and some where there is just an AES block function and the counter is authenticated separately. In general you should probably implement this similarly to how AES-GCM is done, except of course skipping the GCM-specific implementations.

I am not sure why the QUIC spec uses "AES-CTR" specifically since, IIUC, the encryption of the packet number will never be larger than one AES block, so a multi-block mode isn't necessary, AFAICT. Maybe the intent there is to ensure that only the AES encrypt function is needed, and that the AES decrypt function isn't needed. Since the dedicated AES-CTR logic from OpenSSL is optimized to operate on larger-than-single-block inputs, it seems to make sense, for performance, to skip the AES-CTR-specific logic too, and just always use the single-block AES encrypt function.

As far as the API goes, this needs to live outside of the `aead` submodule, of course. I'm not sure exactly what API is optimal. I suggest we get a working PR and then the various people interested in QUIC should try to work together to optimize the API for QUIC's needs.

In particular, if QUIC doesn't actually need to handle multiple blocks, I'd rather have an API that allows for up to but no more than 1 block of output. Later, if we need a more general-purpose AES-CTR API for some other reason, we can add one.

For ChaCha20, I believe that this is mostly a matter of laying a public QUIC-optimized API onto the existing private ChaCha20 API.

**Comment by briansmith** [comment]
Also, we should do the AES and ChaCha20 parts of this in seperate PRs.

**Comment by ghedo** [comment]
@briansmith IIRC the reason why AES-CTR was chosen was to generalize to a certain extent the mechanism for multiple underlying algorithm (so using a stream cipher for both AES and ChaCha) and make implementations easier (and potentially make it easier to add new cipher in the future if the need arises).

@djc We might want to have a generic "StreamingKey" kind of structure for AES128, AES256 and ChaCha, just like the way OpeningKey and SealingKey work, with "init" (so AES initialization can be done only once) and "xor_keystream" methods or something, and a public function like "unathenticated_xor_keystream" that takes the StreamingKey parameter along with an input buffer, and output buffer, a nonce and a counter, that outputs the result of applying the relevant underlying cipher to the input buffer into the output buffer. Though I'll defer to @briansmith to decide whether this makes sense or not :)

I made a rough patch when opening this ticket with only ChaCha support, which basically simply exposes the `chacha20_xor_inner` function through this kind of generic API (though my patch exposed both encrypt and decrypt operations, which we really don't need), and that seemed to work reasonably well, at least in my implementation, but I'm happy for you to do the work and review the code if needed.

**Comment by djc** [comment]
@ghedo I'm very happy for any work you can do on this, I have more than enough things on my plate! Might be good to submit your ChaCha function as a PR so @briansmith can go over it.

I'm still a bit confused by your mention of not needing a decrypt operation. We'll still need an operation to recover the unencrypted packet number from incoming packets, right?

**Comment by ghedo** [comment]
Encryption and decryption is the same operation with stream ciphers (which basically involves XORing the keystream with whatever input you get, either ciphertext or plaintext). So we don't really need separate functions for that.

Anyway, I'll clean up my code a bit and make a PR later.

**Comment by ghedo** [comment]
I opened #695 with the implementation of the generic stream API as well as support for ChaCha20, the main quirk of the API is that in order to support the QUIC case both ChaCha20 and AES need to use 128 bits nonces (in the case of ChaCha it's basically the normal 96 bits nonce with an explicit counter).

I also managed to get the AES-CTR part tto work, though i'll wait for the other PR to be merged to create a new PR, but you can look at the code and test it from https://github.com/ghedo/ring/tree/unauthenticated_stream_aes which has both commits.

**Comment by briansmith** [comment]
I'm closing this. Although we didn't add a general interface for AES-CTR and ChaCha20, we did add an interface for what QUIC implementations need that was even better, so the actual goal of the reporter was met. Thanks for working on this!
